tachyon/crypto/commitments/fri/two_adic_fri_prover.h

```diff
@@ -0,0 +1,167 @@
+#include "tachyon/math/matrix/matrix_types.h"
+#include "tachyon/zk/air/plonky3/challenger/challenger.h"
+
+namespace tachyon::crypto {
+
+template <typename ExtF, typename ChallengeMMCS, typename Challenger>
+CommitPhaseResult<ChallengeMMCS> CommitPhase(
+    TwoAdicFriConfig<ChallengeMMCS>& config,
+    std::vector<std::vector<ExtF>>&& inputs, Challenger& challenger) {
+  // NOTE(ashjeong): This is empirically determined in case the size of the for
+  // loop is small enough to not need parallelization.
+  constexpr size_t kThreshold = 1024;
+
+  using Commitment = typename ChallengeMMCS::Commitment;
+  using ProverData = typename ChallengeMMCS::ProverData;
+
+  // NOTE(ashjeong): Refer to the note below as to why this is safe.
+  std::vector<ExtF> folded = std::move(inputs[0]);
+  size_t total_folds = base::bits::CheckedLog2(folded.size()) -
+                       base::bits::Log2Floor(config.Blowup());
+  std::vector<Commitment> commits;
+  commits.reserve(total_folds);
+  std::vector<ProverData> data;
+  data.reserve(total_folds);
+
+  while (folded.size() > config.Blowup()) {
+    math::RowMajorMatrix<ExtF> leaves((folded.size() + 1) / 2, 2);
+    base::Parallelize(
+        folded,
+        [&leaves](absl::Span<ExtF> chunk, size_t chunk_offset,
+                  size_t chunk_size) {
+          size_t chunk_start = chunk_offset * chunk_size;
+          for (size_t i = chunk_start; i < chunk_start + chunk.size(); ++i) {
+            leaves(i / 2, i % 2) = std::move(chunk[i - chunk_start]);
+          }
+        },
+        kThreshold);
+
+    Commitment commit;
+    ProverData prover_data;
+    CHECK(config.mmcs.Commit({std::move(leaves)}, &commit, &prover_data));
+    commits.push_back(std::move(commit));
+    data.push_back(std::move(prover_data));
+
+    challenger.ObserveContainer(commits.back());
+    const ExtF beta = challenger.template SampleExtElement<ExtF>();
+    VLOG(2) << "FRI(beta[" << commits.size() - 1
+            << "]): " << beta.ToHexString(true);
+
+    folded = FoldMatrix(beta, config.mmcs.GetMatrices(data.back()).back());
+    // NOTE(ashjeong): |inputs| is sorted by largest to smallest size, and the
+    // size of |folded| is divided by two every loop. This means that if
+    // |folded| is intialized as |inputs[0]|, the size of the next iteration of
+    // |folded| will never be the size of |inputs[0]|.
+    for (size_t i = 1; i < inputs.size(); ++i) {
+      if (inputs[i].size() == folded.size()) {
+        base::Parallelize(inputs[i],
+                          [&folded](absl::Span<ExtF> chunk, size_t chunk_offset,
+                                    size_t chunk_size) {
+                            size_t chunk_start = chunk_offset * chunk_size;
+                            for (size_t j = 0; j < chunk.size(); ++j) {
+                              folded[chunk_start + j] += chunk[j];
+                            }
+                          });
```

Comment:
In this case, `OPENMP_PARALLEL_FOR` seems better solution!

Reply:
Do you mean to revert back to before https://github.com/kroma-network/tachyon/pull/492#discussion_r1695225804?

Reply:
Did you use OPENMP_PARALLEL_FOR  before? If so , yes

Reply:
Ah just realized the linked comment was referring to something else. I'll fix to `OPENMP_PARALLEL_FOR`